### ISSUE TITLE
Language plugins: no diagnostic when a config references a plugin's language but omits `language`

### ISSUE THREAD

**Issue by mike-north** [eslint/eslint#21170]
# Language-plugin RFC

## What problem does this feature solve?

A flat-config object that references a language-owning plugin (`plugins: { sh }`, where
`sh.languages.sh` exists) but forgets the `language: "sh/sh"` selector key produces two
confusing outcomes, neither of which names the actual mistake:

1. **Silent fallback, misleading error.** ESLint falls back to the default `"@/js"`
   language and tries to parse the non-JS file as JavaScript. For most non-JS source this
   throws a `Parsing error: Unexpected token ...` fatal lint message that never mentions
   `language`, the offending plugin, or that a language even exists to select.
2. **Hard crash pre-parse.** If the config object also sets `languageOptions` using keys
   that belong to the plugin's language (e.g. `languageOptions: { dialect: 'bash' }`),
   `Config`'s constructor (`lib/config/config.js`) validates those options against the
   **default JS language's** own schema before any plugin code ever runs, and throws
   `TypeError: Key "languageOptions": Unexpected key "dialect" found.` This throw happens
   synchronously inside `FlatConfigArray#getConfig` → `Config`'s constructor, is never
   caught by `Linter#verify`, `ESLint#lintFiles`, or any of the `eslint` CLI's own code, and
   propagates all the way to the CLI's top-level `main().catch(onFatalError)` — producing
   the generic "Oops! Something went wrong!" crash banner (`bin/eslint.js`) instead of a
   normal per-file lint message.

We hit this repeatedly in a DX study for a shell-script language plugin
(`@eslint-sh/eslint-plugin` / `sh-ast`): omitting `language: 'sh/sh'` was the single worst
first-run failure mode, and nothing in either error names the missing key. Investigation
(https://github.com/mike-north/eslint-sh/issues/63) confirmed **this is not something a
language plugin can detect or improve itself** — in both failure modes, ESLint's own core
code (the default JS `Language`'s `parse()`, or `Config`'s own `languageOptions` validation
against the default JS language) runs and either fails or throws *before* any plugin-owned
code path (`Language.parse`, `Language.validateLanguageOptions`, a rule's `create()`) is
ever reached. There is no `Language`/`Plugin` lifecycle hook that fires while a config
object referencing a plugin is being resolved but that plugin's language hasn't been
selected.

## What did you expect to happen?

At minimum, some actionable diagnostic naming the missing `language` key and candidate
plugin language ids, produced by ESLint core (since plugins cannot reach this codepath
themselves).

## What actually happened?

See the two failure modes above. Neither error mentions `language`, the plugin, or the
fact that a custom language exists and wasn't selected.

## Repro

```js
// eslint.config.mjs — using any language-owning plugin, e.g. our sh plugin:
import sh from 'eslint-sh';
export default [
  // missing `language: 'sh/sh'`
  { files: ['**/*.sh'], plugins: { sh }, languageOptions: { dialect: 'bash' } },
];
```

```
TypeError: Key "languageOptions": Unexpected key "dialect" found.
    at new Config (.../eslint/lib/config/config.js:506:10)
    at [finalizeConfig] (.../eslint/lib/config/flat-config-array.js:212:10)
    at FlatConfigArray.getConfigWithStatus (.../@eslint/config-array/dist/cjs/index.cjs:1411:55)
    at FlatConfigArray.getConfig (.../@eslint/config-array/dist/cjs/index.cjs:1429:15)
    at Linter._verifyWithFlatConfigArray (.../eslint/lib/linter/linter.js:1378:30)
```

Uncaught by every layer above it, so it surfaces via the CLI's top-level
`main().catch(onFatalError)` (`bin/eslint.js`) as the generic crash banner rather than a
normal lint result.

## Proposal (concrete, implementable without a new plugin hook)

Both changes are contained to `Config`'s constructor in `lib/config/config.js`, which
already has `plugins` and the resolved `languageOptions` in scope:

1. **Name the missing key's owner in the existing crash.** When the default JS language's
   `validateLanguageOptions` throws `Unexpected key "<key>"`, before rethrowing, probe the
   `plugins` map: for every plugin exposing a `languages` map, check whether that
   language's own `validateLanguageOptions(languageOptions)` accepts the same options
   object without throwing. If exactly one candidate accepts it, append a hint to the
   thrown `TypeError`'s message, e.g. `... Unexpected key "dialect" found. (Plugin "sh"
   defines a language "sh/sh" that accepts this option — did you forget \`language:
   "sh/sh"\` in this config object?)`.
2. **Warn on silent fallback.** When a config object sets `plugins` whose value(s) expose a
   non-empty `languages` map, but the object itself does not set an explicit `language` key
   (so the default `"@/js"` language is used), emit a config-time warning (or a low-severity
   lint message, consistent with how ESLint reports other config-shape issues) naming the
   available plugin language ids, e.g. `Config object matching these files declares
   plugin(s) with custom languages ('sh/sh') but does not select one via \`language\`;
   falling back to the default JavaScript language.`

Both are additive, backward compatible (no schema/behavior change for configs that already
work), and don't require inventing a new `Language`/`Plugin` lifecycle hook — they only use
information `Config` already has when constructing itself.

## Participation

- [x] I am willing to submit a pull request for this change.
